CommonTools/ParticleFlow/plugins/PFCandidateRecalibrator.cc

```diff
@@ -0,0 +1,326 @@
+{
+}
+
+void PFCandidateRecalibrator::produce(edm::Event &iEvent, const edm::EventSetup &iSetup)
+{
+    //access calogeometry
+    edm::ESHandle<CaloGeometry> calogeom;
+    iSetup.get<CaloGeometryRecord>().get(calogeom);
+    const CaloGeometry* cgeo = calogeom.product();
+    const HcalGeometry* hgeom = static_cast<const HcalGeometry*>(cgeo->getSubdetectorGeometry(DetId::Hcal,HcalForward));
+ 
+    //access PFCandidates
+    edm::Handle<reco::PFCandidateCollection> pfcandidates;
+    iEvent.getByToken(pfcandidates_, pfcandidates);
+
+    int nPfCand = pfcandidates->size();
+    std::unique_ptr<reco::PFCandidateCollection> copy(new reco::PFCandidateCollection());
+    std::unique_ptr<reco::PFCandidateCollection> discarded(new reco::PFCandidateCollection());
+    copy->reserve(nPfCand);
+    std::vector<int> oldToNew(nPfCand), newToOld, badToOld; 
+    newToOld.reserve(nPfCand);
+
+    LogDebug("PFCandidateRecalibrator") << "NEW EV:";
+
+    //loop over PFCandidates
+    int i = -1;
+    for (const reco::PFCandidate &pf : *pfcandidates) 
+      {
+	++i;
+	float absEta = std::abs(pf.eta());
+
+	//deal with HE
+	if( pf.particleId() == reco::PFCandidate::ParticleType::h0 && 
+	    absEta > 1.4  && absEta < 3.)
+	  {
+	    bool toKill = false;
+	    for(auto const& badIt: badChHE_)
+	      if( reco::deltaR2(pf.eta(), pf.phi(), badIt.eta, badIt.phi) < 0.07 )
+		toKill = true;
+	    
+	    
+	    if(toKill)
+	      {
+		discarded->push_back(pf);
+		oldToNew[i] = (-discarded->size());
+		badToOld.push_back(i);
+		continue;
+	      }
+	    else
+	      {
+		copy->push_back(pf);
+		oldToNew[i] = (copy->size());
+		newToOld.push_back(i);
+	      }
+	  }
+	//deal with HF
+	else if(absEta >= 3.)
+	  {
+	    const math::XYZPointF& ecalPoint = pf.positionAtECALEntrance();
+	    GlobalPoint ecalGPoint(ecalPoint.X(),ecalPoint.Y(),ecalPoint.Z());
+	    HcalDetId closestDetId(hgeom->getClosestCell(ecalGPoint));
+	    
+	    if(closestDetId.subdet() == HcalForward)
+	      {
+		HcalDetId hDetId(closestDetId.subdet(),closestDetId.ieta(),closestDetId.iphi(),1); //depth1
+		
+		//raw*calEnergy() is the same as *calEnergy() - no corrections are done for HF
+		float longE  = pf.rawEcalEnergy() + pf.rawHcalEnergy()/2.;  //depth1
+		float shortE = pf.rawHcalEnergy()/2.;                       //depth2
+		
+		float ecalEnergy = pf.rawEcalEnergy();
+		float hcalEnergy = pf.rawHcalEnergy();
+		float totEnergy = ecalEnergy + hcalEnergy;
+		float totEnergyOrig = totEnergy;
+		
+		bool toKill = false;
+			
+		for(auto const& badIt: badChHF_)
+		  {
+		    if ( hDetId.ieta() == badIt.ieta &&
+			 hDetId.iphi() == badIt.iphi )
+		      {
+			LogDebug("PFCandidateRecalibrator") << "==> orig en (tot,H,E): " 
+							    << pf.energy() << " " << pf.rawHcalEnergy() << " " << pf.rawEcalEnergy();
+			if(badIt.depth == 1) //depth1
+			  {
+			    longE *= badIt.ratio;
+			    ecalEnergy = longE - shortE;
```

Comment:
from https://github.com/cms-sw/cmssw/pull/22941#discussion_r182123014
since it got harder to find now because git thinks that comment is outdated

> remind me please if it's OK here to have ecalEnergy < 0.
If it has to be non-negative, some truncation logic should be added.

I didn't quite understand the answer.
But maybe my question was unclear.
From https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFProducer/src/PFAlgo.cc#L1545-L1594
it looks like ecalEnergy should not be negative.
Here it is computed as ```longE - shortE``` which is the same as ```rawEcalEnergy*ratio + rawHcalEnergy/2*(ratio - 1)```. So, if ratio is less than 1 the total can be negative and it probably should not be.
There should be some truncation logic (unless it's obvious that ratio is always at least 1, in which case this should be checked with an assert).

Reply:
for this and depth2, I seem to recall that our discussion converged on addition of checks that the output ecalEnergy remains non-negative.

Reply:
hi slava, I've checked this and the requirement `longE > longFibreThr_` works as an effective truncation so ecalEnergy is never negative for the candidates which are saved. logically it seems to me that the current implementation makes sense..

Reply:
for 280385 I see that the smallest ratio is ```0.137208 for HF (HF -32,27,2)``` and the largest is ``` 2.33246 for HF (HF 40,7,1)```.
Only the ones below 0 can lead to negative energies and they do (based on prntouts).

There are also ```inf for HF (HF 40,39,1)``` and ```inf for HF (HF 40,39,2)```.
I think that there should be a protection against this